Win domain group membership module

jiuka opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
With `win_domain_user` it is possible to as a ADUser to a group, but it's not possible to add build and manage ADGroup hierarchies or manage the group membership independent of a user.

This new module `win_domain_group_membership` has the the same parameters/behaviour as the `win_group_membership` module but for AD users instead.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

win_domain_group_membership

##### ADDITIONAL INFORMATION
Example Play
```paste below
---
- name: win_domain_group_member Example
  hosts: adserver
  gather_facts: no
  tasks:
    - name: Add UserName to GroupName
      win_domain_group_membership:
        name: GroupName
        members:
          - UserName
      register: win_domain_group_membership_return

    - name: "Display win_domain_group_membership return"
      debug:
          msg: '{{ win_domain_group_membership_return }}'
```
Example Output
```
PLAY [win_domain_group_member Example] ***********************************************

TASK [Add UserName to GroupName] *******************************************
[GroupName]
+Member = UserName
changed: [adserver]

TASK [DEBUG] *******************************************************************
ok: [adserver] => {
    "msg": {
        "added": [
            "UserName"
        ],
        "changed": true,
        "diff": {
            "prepared": "[GroupName]\n+Member = UserName\n"
        },
        "failed": false,
        "members": [
            "UserName"
        ],
        "name": "GroupName"
    }
}
```

jiuka committed:
Add new win_domain_group_membership module.

jiuka committed:
Add support for diff mode.

jiuka committed:
Do not assign variable which is never used.

dagwieers commented:
@jborean93 I think this one is quite confusing with the existing **win_domain_membership** and I think we need to rename that module to **win_domain_host_membership** instead to clear up this confusion.

jborean93 commented:
@dagwieers I don't really agree, these `win_domain_*` modules are already a mouthful in my mind and think the names as they are are ok.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+    Import-Module ActiveDirectory
+}
+catch {
+    Fail-Json $result "Failed to import ActiveDirectory PowerShell module. This module should be run on a domain controller, and the ActiveDirectory module must be available."
```
I would drop the 'should be run on a domain controller' part, you can run these modules on another host that can communicate with a domain controller and is the reason why we have the `domain_server`, `domain_username`, `domain_password` options to support these scenarios.

jiuka replied:
Changed the error message.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+    Fail-Json $result "Failed to import ActiveDirectory PowerShell module. This module should be run on a domain controller, and the ActiveDirectory module must be available."
+}
+
+$ErrorActionPreference = "Stop"
```
This is automatically set since devel/2.8 when executing the module so isn't needed here.

jiuka replied:
removed it

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+
+$result = @{
+    changed = $false
+    name = $name
```
We should avoid behaviour like this. We have a mechanism to report on the module invocation args in 2.8 which is to use the `AnsibleBasic` util as documented at https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development. You don't have to rewrite your PR as it isn't mandatory yet but we should avoid setting return values for things we already know.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+}
+
+try {
+    Get-ADGroup -Identity $name -Properties * @extra_args
```
Do we really need all the properties? I would also loose the try/catch block. Historically we've done this so we can at least give a somewhat pretty error message but in devel/2.8 this is done automatically for us. Only catch exceptions if we need to handle specific behaviour or whether we could be more explicit in an exception based on the exception type.

jiuka replied:
I removed the `try`/`catch` around the Get-AD* and let ansible handle the exceptions. The whole test if the group exists in the first place has become obsolete with this as it will return the same exception the the first time the group is accessed.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+        $group_member = Get-ADUser -Identity $member @extra_args
+    }
+    catch {
+        $group_member = $null
```
I would add a comment stating the reasons why we need the empty catch as the `Get-ADUser` doesn't respect the `-ErrorAction SilentlyContinue` method.

jiuka replied:
Added comments for both `try`/`catch`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,155 @@
+    $result.members = @()
+}
+
+if ($diff_mode -and $result.changed) {
```
Using the `prepared` part of a diff should be avoided. If you use the before/after keys under diff Ansible will generate a nicer diff output with colour, unlike prepared.

Because of this I would do something like

```
$result.diff.before = @(list of members originally in the group)
$result.diff.after = @(list of member now in the group after module execution)
```

The controller side code will handle the + and - elements for you based on the list diffs.

jiuka replied:
A List did not work for me. It complained about not finding split on the list. But changed from `prepared ` to `before`/`after` for the diff.

jborean93 replied:
Hmm ok, I would need to test out that, I've only really done a dict that contained a list with the values and not a list directly. Thanks for letting me know.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.py:
```diff
@@ -0,0 +1,93 @@
+description:
+     - Allows the addition and removal of domain users
+       and domain groups from/to a domain group.
+options:
```
You are missing the `domain_*` module options.

jiuka replied:
Added the documentation for domain_* options

jborean93 reviewed (changes requested):
Looks good for a first attempt, have added some review comments.

jiuka committed:
Add documentation for the `domain_*` options.

jiuka committed:
Let ansible handle the exceptions.

The test if the group exists is useless as the first action on the 
groups fails with the same error message if it does not exist.

jiuka committed:
Add comments why we need the try/catch

jiuka committed:
Rework diff handling.

Just return before/after state and let ansible do the working out of the
diff.

jiuka committed:
Minor cleanups according to PR

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,141 @@
+    # Use try/catch as Get-ADUser does not support `-ErrorAction SilentlyContinue`
+    # We do not want to fail but query for group if no user is found
+    try {
+        $group_member = Get-ADUser -Identity $member @extra_args
```
It might be better to use `Get-ADObject` instead of the specific `Get-ADUser` or `Get-ADObject`. IIRC you can set more than just groups/users as a member of a domain group but I could be wrong here.

jiuka replied:
According to `Add-ADGroupMember` it's `users`, `groups`, `service accounts`, or `computer`. However `Get-ADObject` returns even more Object types and its `-Identity` parameter does only take the `DN ` or `GUID ` of an object so the query has a quiet long filter, however its now down to only one query without try/catch

jborean93 reviewed (commented):
Looks good, just one question around the `Get-ADUser/Get-ADGroup` vs `Get-ADObject`

jiuka committed:
Switch from Get-AdUser/Group to Get-AdObject

so we can add/remove service accounts, or computers too.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    Import-Module ActiveDirectory
+}
+catch {
+    Fail-Json $result "win_domain_group_membership requires the ActiveDirectory PS module to be installed"
```
$result is not defined at this particular point in time and we try and use the parameter names where possible. This can simply be changed to `Fail-Json -obj @{} -message "win_......`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+$members = Get-AnsibleParam -obj $params -name "members" -type "list" -failifempty $true
+
+$extra_args = @{}
+if ($domain_username -ne $null) {
```
Standard PS practice is to have `$null` on the left hand side in equality checks see https://github.com/PowerShell/PSScriptAnalyzer/blob/development/RuleDocumentation/PossibleIncorrectComparisonWithNull.md for more info on this. This will be the same for some other checks you have below where `$null` is on the right.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    changed = $false
+}
+if ($state -in @("present", "pure")) {
+    $result.added = @()
```
It would be a lot easier to just have `added` and `removed` present in the return result regardless of the state. I would also use a generic list instead of an array or ArrayList like so;

```
$result = @{
    changed = $false
    added = [System.Collections.Generic.List`1[String]]@()
    removed = [System.Collections.Generic.List`1[String]]@()
}
```

This is recommend because;

* It less overhead when adding elements to a list compared to an array, an array has to initialise a new array of a larger size, copy the previous elements to the new array then add the new value. A list just has to add the new element to an existing list.
* Some people just use `[System.Collections.ArrayList]`, a list is recommended because the type is specified so we can guarantee it's a list of Strings, we don't have to deal with object boxing/unboxing and the `.Add()` method for a List does not return the index whereas ArrayList's one does

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    $result.added = @()
+}
+if ($state -in @("absent", "pure")) {
+    $result.removed = @()
```
See comment on `added`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+# Module control parameters
+$state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "present","absent","pure"
+$domain_username = Get-AnsibleParam -obj $params -name "domain_username" -type "str"
+$domain_password = Get-AnsibleParam -obj $params -name "domain_password" -type "str" -failifempty ($domain_username -ne $null)
```
($null -ne $domain_username)

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    $credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $domain_username, $domain_password
+    $extra_args.Credential = $credential
+}
+if ($domain_server -ne $null) {
```
($null -ne $domain_server)

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+$pure_members = @()
+
+foreach ($member in $members) {
+    $group_member = Get-ADObject -Filter "SamAccountName -eq '$member' -and (ObjectClass -eq 'user' -or ObjectClass -eq 'group' -or ObjectClass -eq 'computer' -or ObjectClass -eq 'msDS-ManagedServiceAccount')"
```
You need to specify the properties you want to compare and use, e.e. this should have `-Properties objectSid, sAMAccountName` so you can use it below. This line is also pretty long, it might be better to define the filter in a separate string and just pass that into the `-Filter` parameter to make it easier to read.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+}
+
+$members_before = Get-AdGroupMember -Identity $name @extra_args
+$pure_members = @()
```
Use 

```
$pure_members = [System.Collections.Generic.List`1[Object]]@()
```

I'm not sure what the object that it will be encapsulating but for we probably don't really care about that too much here and using the `Object` type is fine.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    if ($state -eq "pure") {
+        $pure_members += $group_member
```
Once `$pure_members` has been changed, this can be `$pure_members.Add($group_member)`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+
+    $user_in_group = $false
+    foreach ($current_member in $members_before) {
+        if ($current_member.sid -eq $group_member.sid) {
```
This might be a side effect of moving to `Get-ADObject` but `sid` is not a valid property of the return object. Once you add the `-Properties objectSid` to the `Get-ADObject` cmdlet call, this should be `if ($current_member.sid -eq $group_member.objectSid)`. I found that without this the module was no longer idempotent because it never matched with the input accounts.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+
+    if ($state -in @("present", "pure") -and !$user_in_group) {
+        Add-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args
+        $result.added += $group_member.SamAccountName
```
With the array to List this change this would be `$result.added.Add($group_member.sAMAccountName)`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+        $result.changed = $true
+    } elseif ($state -eq "absent" -and $user_in_group) {
+        Remove-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args -Confirm:$False
+        $result.removed += $group_member.SamAccountName
```
With the array to List this change this would be `$result.removed.Add($group_member.sAMAccountName)`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    foreach ($current_member in $current_members) {
+        $user_to_remove = $true
+        foreach ($pure_member in $pure_members) {
+            if ($pure_member.sid -eq $current_member.sid) {
```
Same thing here, the comparison should be `if ($pure_member.objectSid -eq $current_member.sid) {`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+
+        if ($user_to_remove) {
+            Remove-ADGroupMember -Identity $name -Members $current_member -WhatIf:$check_mode @extra_args -Confirm:$False
+            $result.removed += $current_member.SamAccountName
```
`$result.removed.Add($current_member.sAMAccountName)`

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    if (!$check_mode) {
+        $result.diff.after = [Array]$final_members.SamAccountName | Out-String
+    } else {
+        $after = [System.Collections.ArrayList][Array]$final_members.SamAccountName
```
The use of `ArrayList` here means that `Add` is returning the index where the object was added at. This can cause issues when Ansible goes to parse the return json from a module. Unfortunately the Remove method of a List does return a bool but we can still handle that.

I would recommend you use a List like

```
$after = [System.Collections.Generic.List`1[String]]$final_members.sAMAccountName
```

There should also be no reason why you need the `[Array]` cast as well.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+        $result.diff.after = [Array]$final_members.SamAccountName | Out-String
+    } else {
+        $after = [System.Collections.ArrayList][Array]$final_members.SamAccountName
+        $result.removed | ForEach-Object { $after.Remove($_) }
```
Add `$after.Remove($_) > $null` to make sure we are discarding the return value and not outputting it to the Ansible controller.

jborean93 commented on lib/ansible/modules/windows/win_domain_group_membership.ps1:
```diff
@@ -0,0 +1,125 @@
+    } else {
+        $after = [System.Collections.ArrayList][Array]$final_members.SamAccountName
+        $result.removed | ForEach-Object { $after.Remove($_) }
+        $result.added | ForEach-Object { $after.Add($_) }
```
Easier to just do `$after.AddRange($result.added)`

jborean93 reviewed (changes requested):
I did some testing and it seems like the module is no longer idempotent. Have added some recommendations to fix that back up.

I also like how you dealt with the list issue with before and after, very ingenious.

jiuka committed:
Cleanup PowerShell code

jborean93 reviewed (approved):
Looks great, thanks for working on this one.
